pkg/apis/crd/v1beta1/types.go

```diff
@@ -151,3 +151,267 @@ type ControllerCondition struct {
+	ComponentEgress        TraceflowComponent = "Egress"
+)
+
+type TraceflowAction string
+
+const (
+	ActionDelivered TraceflowAction = "Delivered"
+	ActionReceived  TraceflowAction = "Received"
+	ActionForwarded TraceflowAction = "Forwarded"
+	ActionDropped   TraceflowAction = "Dropped"
+	ActionRejected  TraceflowAction = "Rejected"
+	// ActionForwardedOutOfOverlay indicates that the packet has been forwarded out of the network
+	// managed by Antrea. This indicates that the Traceflow request can be considered complete.
+	ActionForwardedOutOfOverlay TraceflowAction = "ForwardedOutOfOverlay"
+	ActionMarkedForSNAT         TraceflowAction = "MarkedForSNAT"
+	ActionForwardedToEgressNode TraceflowAction = "ForwardedToEgressNode"
+)
+
+// List the supported protocols and their codes in traceflow.
+// According to code in Antrea agent and controller, default protocol is ICMP if protocol is not inputted by users.
+const (
+	ICMPProtocolNumber int32 = 1
+	IGMPProtocolNumber int32 = 2
+	TCPProtocolNumber  int32 = 6
+	UDPProtocolNumber  int32 = 17
+	SCTPProtocolNumber int32 = 132
+)
+
+var SupportedProtocols = map[string]int32{
+	"TCP":  TCPProtocolNumber,
+	"UDP":  UDPProtocolNumber,
+	"ICMP": ICMPProtocolNumber,
+}
+
+var ProtocolsToString = map[int32]string{
+	TCPProtocolNumber:  "TCP",
+	UDPProtocolNumber:  "UDP",
+	ICMPProtocolNumber: "ICMP",
+	IGMPProtocolNumber: "IGMP",
+	SCTPProtocolNumber: "SCTP",
+}
+
+// List the supported destination types in traceflow.
+const (
+	DstTypePod     = "Pod"
+	DstTypeService = "Service"
+	DstTypeIPv4    = "IPv4"
+)
+
+var SupportedDestinationTypes = []string{
+	DstTypePod,
+	DstTypeService,
+	DstTypeIPv4,
+}
+
+// List the ethernet types.
+const (
+	EtherTypeIPv4 uint16 = 0x0800
+	EtherTypeIPv6 uint16 = 0x86DD
+)
+
+// Default timeout in seconds.
+const DefaultTraceflowTimeout uint16 = 20
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TraceflowSpec   `json:"spec,omitempty"`
+	Status TraceflowStatus `json:"status,omitempty"`
+}
+
+// TraceflowSpec describes the spec of the traceflow.
+type TraceflowSpec struct {
+	Source      Source      `json:"source,omitempty"`
+	Destination Destination `json:"destination,omitempty"`
+	Packet      Packet      `json:"packet,omitempty"`
+	// LiveTraffic indicates the Traceflow is to trace the live traffic
+	// rather than an injected packet, when set to true. The first packet of
+	// the first connection that matches the packet spec will be traced.
+	LiveTraffic bool `json:"liveTraffic,omitempty"`
+	// DroppedOnly indicates only the dropped packet should be captured in a
+	// live-traffic Traceflow.
+	DroppedOnly bool `json:"droppedOnly,omitempty"`
+	// Timeout specifies the timeout of the Traceflow in seconds. Defaults
+	// to 20 seconds if not set.
+	Timeout uint16 `json:"timeout,omitempty"`
```

Comment:
The K8s API design guide actually discourages the use of unsigned integers:

> Do not use unsigned integers, due to inconsistent support across languages and libraries. Just validate that the integer is non-negative if that's the case.

Source: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

As a rule I would say we should never use unsigned integers in API. There are only a few scenarios in which it may make sense, for example for packet field values, but even then I see that we use `int32` for the IP protocol.

Note that the same document recommends using int32 / int64. I don't see int16s in the core K8s API. I feel like we could make this field an int32.

This change should also be mentioned in the commit message / PR description, once it is implemented.

Reply:
Done, message updated.